### PR TITLE
(WIP) Add Kyber Balancer bridge reserve

### PR DESCRIPTION
**Pull request by manhlx3006**

### CODE REVIEW

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol [hunk 1]
```diff
@@ -0,0 +1,253 @@
+pragma solidity 0.6.6;
+pragma experimental ABIEncoderV2;
+
+import "./IBalancerExchangeProxy.sol";
+import "../../IKyberReserve.sol";
+import "../../IERC20.sol";
+import "../../utils/Withdrawable3.sol";
+import "../../utils/Utils5.sol";
```
Comment (ilanDoron) [inline contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:8]: please use utils from NPM package.

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    ) external override view returns (uint256 rate) {
+        if (!tradeEnabled) return 0;
+        if (srcQty == 0) return 0;
+
+        uint256 destQty;
+        if (src == ETH_TOKEN_ADDRESS) {
+            uint256 nPools = numberPools[dest];
+            if (nPools == 0) return 0; // token is not listed
+            (, destQty) = exchangeProxy.viewSplitExactIn(
+                weth,
```
Comment (ilanDoron) [inline contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:207]: here should use weth while in trade can use eth (0xeeee....?)
not clear why this diff.

**Review by ilanDoron — COMMENTED** [review]
see some small comments

**Review by ilanDoron — COMMENTED** [review]
please share some data regarding swaps
maybe in ropsten?
or mainnet if u have them

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol (hunk 2 above)
Comment (manhlx3006) [reply contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:207]: yes this is how it works, the viewSplitExactIn func only accepts weth address, but the swap function supports eth addres

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            srcToken,
+            destToken,
+            srcAmount,
+            conversionRate
+        );
+
+        uint256 value;
+        uint256 nPools;
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount, "msg.value != srcAmount");
```
Comment (Anyhowclick) [inline contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:87]: proxy does this check already right?

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool /* validate */
+    ) external override payable returns (bool) {
+        require(tradeEnabled, "trade is disabled");
+        require(msg.sender == kyberNetwork, "only kyberNetwork");
+
+        require(conversionRate > 0, "conversionRate 0");
```
Comment (Anyhowclick) [inline contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:76]: weeelll if we trust the network to give valid rate, this won't happen right =)

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        );
+
+        uint256 value;
+        uint256 nPools;
+        if (srcToken == ETH_TOKEN_ADDRESS) {
+            require(msg.value == srcAmount, "msg.value != srcAmount");
+            value = srcAmount;
+            nPools = numberPools[destToken];
+        } else {
+            require(msg.value == 0, "msg.value is not 0");
```
Comment (Anyhowclick) [inline contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:91]: here too

File: contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol (hunk 4 above)
Comment (manhlx3006) [reply contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol:76]: not totally trust, you can see that we always verify destAmount >= expectedDestAmount
but in scope of bridge reserves, we can safely remove unnecessary check

**Review by ilanDoron — COMMENTED** [review]
I agree with Desmond we can remove some safety checks.
and also move this code to new bridge repository.

**Comment by ilanDoron** [comment]
code moving to another repository.

**Pull request closed without merging**